aptos-move/framework/aptos-experimental/sources/daa_siws_phantom.move

```diff
@@ -0,0 +1,255 @@
+/// Derivable account abstraction that verifies a message signed by
+/// SIWS from a Phantom wallet.
+module aptos_experimental::daa_siws_phantom {
+    use aptos_framework::auth_data::AbstractionAuthData;
+    use aptos_std::ed25519::{
+        Self,
+        new_signature_from_bytes,
+        new_unvalidated_public_key_from_bytes,
+    };
+    use std::chain_id;
+    use std::error;
+    use std::option::{Self, Option};
+    use std::string_utils;
+    use std::transaction_context::{Self, EntryFunctionPayload};
+    use std::vector;
+
+    /// Signature failed to verify.
+    const EINVALID_SIGNATURE: u64 = 1;
+    /// Non base58 character found in public key.
+    const EINVALID_BASE_58_PUBLIC_KEY: u64 = 2;
+    /// Entry function payload is missing.
+    const EMISSING_ENTRY_FUNCTION_PAYLOAD: u64 = 3;
+
+    const BASE_58_ALPHABET: vector<u8> = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+    const HEX_ALPHABET: vector<u8> = b"0123456789abcdef";
+
+    fun split_abstract_public_key(abstract_public_key: &vector<u8>): (vector<u8>, vector<u8>) {
+        // First 44 bytes are the base58 utf8 encoded public key
+        let base58_public_key = abstract_public_key.slice(0, 44);
+        let domain = abstract_public_key.slice(44, abstract_public_key.length());
+        (base58_public_key, domain)
+    }
+
+    fun network_name(): Option<vector<u8>> {
+        let chain_id = chain_id::get();
+        if (chain_id == 1) {
+            option::some(b"mainnet")
+        } else if (chain_id == 2) {
+            option::some(b"testnet")
+        } else if (chain_id == 4) {
+            option::some(b"local")
+        } else {
+            option::none()
+        }
+    }
+
+    fun construct_message(
+        base58_public_key: &vector<u8>,
+        domain: &vector<u8>,
+        entry_function_name: &vector<u8>,
+        digest_utf8: &vector<u8>,
+    ): vector<u8> {
+        let message = &mut vector[];
+        message.append(*domain);
+        message.append(b" wants you to sign in with your Solana account:\n");
+        message.append(*base58_public_key);
+        message.append(b"\n\nTo execute transaction ");
+        message.append(*entry_function_name);
+        message.append(b" on Aptos blockchain");
+        let maybe_network_name = network_name();
+        if (maybe_network_name.is_some()) {
+            message.append(b" (");
+            message.append(maybe_network_name.destroy_some());
+            message.append(b")");
+        };
```

Comment:
nobody that is not developer will be playing with devnets , so having a `developer network #164` with number there for non mainnet/testnet seems much cleaner